### ISSUE TITLE
Save the `.`

### ISSUE THREAD

**Issue by laktak** [vim/vim#6299]
I often search through text or code and use the `.` command to repeat my change (for example change a name etc.).

Halfway through I notice another change I'd like to make (e.g. remove a misplaced comma, add an s to the end of a word ...) but by making the change my `.` command would be cleared.

So now I either have to interrupt my workflow, make the change and then start again or remember the position to go back later.

I think being able to save the action of the `.` to a register and later repeat it as a macro would solve the problem.

**Comment by inkarkat** [comment]
I really like this idea, and think that this would be a great addition. Here's some brainstorming on how this could be offered:

* If the last change command and applicable scope / motion could be represented in a meaningful way, so that it could also be synthesized by plugins, such functionality could also make any plugin action repeatable, and would allow to natively support what many plugins today use the [repeat.vim plugin](https://www.vim.org/scripts/script.php?script_id=2136) for (which remaps the `.` and `u` keys to hook into the redo system). That would be really great.

* If clients would only be able to treat the value of a hypothetical `getlastchange()` / `setlastchange()` as an opaque blob, it would be far less useful, allowing not much more than what you suggest, though I would love to have a mapping that makes the next normal mode command exempt from being recorded as the last change, too.

* Somewhat in between, if last change and scope could be described, but that representation would not be enough to synthesize new artificial changes (maybe because the internal structure within Vim is complex), a change history could be offered, like `q:` for Ex commands (e.g. via a new `q.` command). This could be used to restore any previous change, covering your use case (of skipping the most recent change) and also allowing access to earlier changes, too. This solution would offer what is described in the previous paragraph more or less automatically through `histget()` / `histadd()`.

**Comment by lacygoill** [comment]
[Item from the todo list](https://github.com/vim/vim/blob/1089374130862e5c19745d3ff4314f0c5690baf1/runtime/doc/todo.txt#L634-L636) which may be relevant:

> Redo only remembers the last change.  Could use "{count}g." to redo an older
> change.  How does the user know which change?  At least have a way to list
> them: ":repeats".  Add to history, like search history and command line history.

**Comment by jlittlenz** [comment]
I've wished for this for a long time, going back to my vi days.

Like there's a scratch register for deleted or changed text, could there be a scratch register that records every change? So macro recording is always on?

**Comment by vim-ml** [comment]
On 2020-06-19, lacygoill wrote:
> Item from the todo list which may be relevant:
> 
>     Redo only remembers the last change. Could use "{count}g." to redo an older
>     change. How does the user know which change? At least have a way to list
>     them: ":repeats". Add to history, like search history and command line
>     history.

I like that.  I had thought of keeping each change in a stack, so
that you could pop the last change off the stack if you wanted the
one before, but I like this better.  It would be especially nice if
"g." (without a count) executed the next-to-last change.

Regards,
Gary

**Comment by brammool** [comment]
> On 2020-06-19, lacygoill wrote:
> > Item from the todo list which may be relevant:
> > 
> >     Redo only remembers the last change. Could use "{count}g." to redo an older
> >     change. How does the user know which change? At least have a way to list
> >     them: ":repeats". Add to history, like search history and command line
> >     history.
> 
> I like that.  I had thought of keeping each change in a stack, so
> that you could pop the last change off the stack if you wanted the
> one before, but I like this better.  It would be especially nice if
> "g." (without a count) executed the next-to-last change.

There are some questions about how this behaves exactly.  E.g. after
using "3g." does that redo operation become the current one, so that "."
will redo it again?  That's a bit like how the command history works.
Then will "3g." do the same thing, or have the entries shifted?

:repeats could list them with the "g.", "2g.", etc. command that would
execute them.

I would find this most useful when going through a list of matches,
either with "n" or ":cn", and seeing something else that needs changing.
If that is only one operation that using "g." would work, but if that
ends up being two or three changes, you get lost.  Perhaps we need a way
to "save the current redo" and "restore the saved redo"?  Or switching
between two redo histories.

-- 
If cars evolved at the same rate as computers have, they'd cost five euro, 
run for a year on a couple of liters of petrol, and explode once a day.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by tejr** [comment]
When I first thought how handy something like this would be, the way I imagined it working would be to automatically store the relevant keys for the previous change in a read-only register.

**Comment by chrisbra** [comment]
I think there even was a patch for accessing the `.` register floating around. Here it is: https://groups.google.com/d/msg/vim_dev/TIH6wsD4Qo4/msUMWHnWt3oJ

**Comment by laktak** [comment]
> There are some questions about how this behaves exactly. E.g. after using "3g." does that redo operation become the current one, so that "." will redo it again? That's a bit like how the command history works. Then will "3g." do the same thing, or have the entries shifted?

Yes, I would expect it to become the current one, so I can continue using it with `.` (especially when going through matches).

Just having `g.` would be great because you still remember what that does. I'm not sure if I would find a `:repeats` as helpful ... maybe for the case where you do multiple changes before you go to the next match.

**Comment by tejr** [comment]
Thank you @chrisbra .  I really like this idea and would love to see it in mainline Vim.

**Comment by Shane-XB-Qian** [comment]
> :repeats could list them with the "g.", "2g.", etc. command that would execute them.

@brammool maybe [also] working with 'g,', 'g;' would be better, enhance it: echo the last 'repeat' at that location, then 'g.' switch to it, now '.' can repeat that one ..

**Comment by Houl** [comment]
I use `g.` to repeat `n.` on the Visual area or the remaining buffer.  No, it's not released anywhere, pretty sure someone has done the same (in a more polished manner) (using which key combo?).  Just want to point out a conflicting use ... which IMHO is even more useful, and also is a candidate for a builtin command.

**Comment by laktak** [comment]
@chrisbra thanks for the link! I've updated it, added `setreg()` support and created a PR. Could you take a look at it? #6346

**Comment by Shane-XB-Qian** [comment]
jump-in .. :)
um, what was that gg link saying? (no chance to open gg link -_-#)
i was thinking 'g.' working with 'g,' and 'g;', but now - your pr, if so, no 2,3,4..g. but only 2, then how or was there still 'g.' to switch ?

**Comment by sheerun** [comment]
I think it would be really really nice if vim implemented functionality of `repeat.vim` natively (so developers can repeat not only last command, but last group of commands), I can't use vim without `repeat.vim`. Macros are nice but hard to use, and you need to remember to start recording.
